dpe/src/x509.rs

```diff
@@ -2227,6 +2235,286 @@ impl CertWriter<'_> {
+}
+
+fn get_tci_nodes(
+    dpe: &mut DpeInstance,
+    handle: &ContextHandle,
+    locality: u32,
+    nodes: &mut [TciNodeData],
+) -> Result<usize, DpeErrorCode> {
+    let parent_idx = dpe.get_active_context_pos(handle, locality)?;
+    let tcb_count = dpe.get_tcb_nodes(parent_idx, nodes)?;
+    if tcb_count > MAX_HANDLES {
+        return Err(DpeErrorCode::InternalError);
+    }
+    Ok(tcb_count)
+}
+
+fn get_subject_key_identifier(
+    env: &mut DpeEnv<impl DpeTypes>,
+    pub_key: &EcdsaPub,
+    subject_key_identifier: &mut [u8],
+) -> Result<(), DpeErrorCode> {
+    // compute key identifier as SHA hash of the DER encoded subject public key
+    let mut hasher = env.crypto.hash_initialize(DPE_PROFILE.alg_len())?;
+    hasher.update(&[0x04])?;
+    hasher.update(pub_key.x.bytes())?;
+    hasher.update(pub_key.y.bytes())?;
+    let hashed_pub_key = hasher.finish()?;
+    if hashed_pub_key.len() < MAX_KEY_IDENTIFIER_SIZE {
+        return Err(DpeErrorCode::InternalError);
+    }
+    // truncate key identifier to 20 bytes
+    subject_key_identifier.copy_from_slice(&hashed_pub_key.bytes()[..MAX_KEY_IDENTIFIER_SIZE]);
+    Ok(())
+}
+
+// TODO(clundin): Remove this lint when adding the DeriveContext certificate generation code.
+#[allow(unused)]
+pub(crate) fn create_exported_dpe_cert(
+    args: &CreateDpeCertArgs,
+    dpe: &mut DpeInstance,
+    env: &mut DpeEnv<impl DpeTypes>,
+    cert: &mut [u8],
+) -> Result<CreateDpeCertResult, DpeErrorCode> {
+    create_dpe_cert_or_csr(
+        args,
+        dpe,
+        env,
+        CertificateFormat::X509,
+        CertificateType::Exported,
+        cert,
+    )
+}
+
+pub(crate) fn create_dpe_cert(
+    args: &CreateDpeCertArgs,
+    dpe: &mut DpeInstance,
+    env: &mut DpeEnv<impl DpeTypes>,
+    cert: &mut [u8],
+) -> Result<CreateDpeCertResult, DpeErrorCode> {
+    create_dpe_cert_or_csr(
+        args,
+        dpe,
+        env,
+        CertificateFormat::X509,
+        CertificateType::Leaf,
+        cert,
+    )
+}
+
+pub(crate) fn create_dpe_csr(
+    args: &CreateDpeCertArgs,
+    dpe: &mut DpeInstance,
+    env: &mut DpeEnv<impl DpeTypes>,
+    csr: &mut [u8],
+) -> Result<CreateDpeCertResult, DpeErrorCode> {
+    create_dpe_cert_or_csr(
+        args,
+        dpe,
+        env,
+        CertificateFormat::Csr,
+        CertificateType::Leaf,
+        csr,
+    )
+}
+
+fn create_dpe_cert_or_csr(
+    args: &CreateDpeCertArgs,
+    dpe: &mut DpeInstance,
+    env: &mut DpeEnv<impl DpeTypes>,
+    cert_format: CertificateFormat,
+    cert_type: CertificateType,
+    output_cert_or_csr: &mut [u8],
+) -> Result<CreateDpeCertResult, DpeErrorCode> {
+    let algs = DPE_PROFILE.alg_len();
+    let digest = get_dpe_measurement_digest(dpe, env, args.handle, args.locality)?;
+
+    let key_pair = match cert_type {
+        CertificateType::Exported => {
+            let cdi = env
+                .crypto
+                .derive_exported_cdi(algs, &digest, args.cdi_label)?;
```

Comment:
nit: I wonder if the key derivation code should stay in the CertifyKey command implementation. Since it's not really related to X.509.

Reply:
I agree but I could not think of a way to do it.

Since they are an associated type of `DpeEnv<impl DpeTypes>` I don't think there is a type that we can use in the function signature.

Any ideas?

Reply:
The CDI is an associated type, but the key pair derivation functions return a raw public key that we can just pass into this function, right?

Reply:
I think the private key is still needed for the exported certificate since it's self signed.

This has made me realized that when I refactored the code, I did not self-sign the exported certificate!

Reply:
As discussed offlne, the cert should still be signed by the alias key